src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+{
+    auto assetPair = std::make_pair(selling, buying);
+    auto iter = mLoadBestOfferContext.find(assetPair);
+    if (iter == mLoadBestOfferContext.end())
+    {
+        auto res = mLoadBestOfferContext.insert(std::make_pair(
+            assetPair, LoadBestOfferContext(selling, buying, *this)));
+        iter = res.first;
+    }
+    return iter->second;
+}
+
+void
+LedgerState::invalidateLoadBestOfferContext(Asset const& selling,
+                                            Asset const& buying)
+{
+    auto assetPair = std::make_pair(selling, buying);
+    auto iter = mLoadBestOfferContext.find(assetPair);
+    if (iter != mLoadBestOfferContext.end())
+    {
+        mLoadBestOfferContext.erase(iter);
+    }
+}
+
+std::vector<LedgerState::StateEntry>
+LedgerState::getOffers(Asset const& selling, Asset const& buying)
+{
+    std::vector<StateEntry> offers;
+    std::set<LedgerKey> seen;
+    getOffers(selling, buying, offers, seen);
+    return offers;
+}
+
+void
+LedgerState::getOffers(Asset const& selling, Asset const& buying,
+                       std::vector<StateEntry>& offers,
+                       std::set<LedgerKey>& seen)
+{
+    for (auto const& state : mState)
+    {
+        if (state.first.type() != OFFER)
+        {
+            continue;
+        }
+        else if (!seen.insert(state.first).second)
+        {
+            continue;
+        }
+        else if (!state.second->ignoreInvalid().entry())
+        {
+            continue;
+        }
+        assert(!state.second->valid());
+
+        // Note: Can't compare Assets with !=
+        auto const& offer = state.second->ignoreInvalid().entry()->data.offer();
+        if (!(selling == offer.selling) || !(buying == offer.buying))
+        {
+            continue;
+        }
+        offers.push_back(state.second);
+    }
+
+    if (mParent)
+    {
+        mParent->getOffers(selling, buying, offers, seen);
+    }
+}
+
+LedgerState::LoadBestOfferContext::LoadBestOfferContext(
+    Asset const& selling, Asset const& buying, LedgerState& ledgerState)
+    : mSelling(selling)
+    , mBuying(buying)
+    , mLedgerState(ledgerState)
+    , mTop(nullptr)
+    , mInMemory([](StateEntry const& lhs,
+                   StateEntry const& rhs) { return compareOffers(rhs, lhs); },
+                mLedgerState.getOffers(selling, buying))
+    , mOffersLoadedFromDatabase(0)
+{
+}
+
+void
+LedgerState::LoadBestOfferContext::loadFromDatabaseIfNecessary()
+{
+    if (mFromDatabase.empty())
+    {
+        LedgerState* root = &mLedgerState;
+        while (root->mParent != nullptr)
+        {
+            root = root->mParent;
+        }
+        mFromDatabase = root->loadBestOffersFromDatabase(
+            5, mOffersLoadedFromDatabase, mSelling, mBuying);
+        mOffersLoadedFromDatabase += mFromDatabase.size();
+    }
+}
+
+LedgerState::StateEntry
+LedgerState::LoadBestOfferContext::loadBestOffer()
+{
+    if (mTop)
+    {
+        assert(!mTop->valid());
```

Comment:
I find that counter intuitive that we are adding on purpose non-valid item to mInMemory here.
Maybe this is a problem with name of "valid" method?

Reply:
I agree that this could indicate a problem with the name.

`LoadBestOfferContext` maintains the invariant that there is at most one valid `LedgerEntryReference` referring to an offer (with the correct assets), and if there is one then it is the one that was just returned by `loadBestOffer`. This means it is incorrect to call `loadBestOffer` if any `LedgerEntryReference` referring to an offer (with the correct assets) is valid.

Why is this required though? I think it would be possible to implement this without this constraint, but it would make it impossible to do the performance optimization which is done in `LoadBestOfferContext`. Specifically, if there could be multiple valid `LedgerEntryReferences` referring to offers (with the correct assets) then `LoadBestOfferContext::loadBestOffer` would have to re-process the entire `mInMemory` on each invocation. It is easy to see that this is the case. Any valid `LedgerEntryReference` can be modified at any time (and there is no callback to notify `LedgerState` that a modification has occurred). Suppose that I invoke `LoadBestOfferContext::loadBestOffer`, then modify many offers (with the correct assets), then call `LoadBestOfferContext::loadBestOffer` again. At this point `mInMemory` could be invalid (in the heap sense) at arbitrarily many locations, so it would need to be re-formed into a heap. But that would be no better than just iterating to find the best offer at every invocation. By making the restriction that if a `LedgerEntryReference` referring to an offer (with the correct assets) is only valid if it was just returned by `loadBestOffer` then we know that modifications can only occur at the top of the heap.

The consequence of this is that a sequence of calls to `loadBestOffer` can be efficient if you do not make any intervening calls to `LedgerState::load` or `LedgerState::create` that constructed a `LedgerEntryReference` referring to an offer (with the correct assets). If such an intervening call is made, then the `LoadBestOfferContext` is destroyed so that it has to be reconstructed at the next call to `loadBestOffer`, thereby maintaining the constraint.